crates/typst-utils/src/format.rs

```diff
@@ -0,0 +1,139 @@
+use std::{cmp::min, time::Duration};
+
+/// Returns number of `(days, hours, minutes, seconds, milliseconds, microseconds)`.
+fn get_duration_parts(duration: &Duration) -> (u16, u8, u8, u8, u16, u16) {
+    // In practice we probably don't need nanoseconds.
+    let micros = duration.as_micros();
+    let (millis, micros) = (micros / 1000, (micros % 1000) as u16);
+    let (sec, millis) = (millis / 1000, (millis % 1000) as u16);
+    let (mins, sec) = (sec / 60, (sec % 60) as u8);
+    let (hours, mins) = (mins / 60, (mins % 60) as u8);
+    let (days, hours) = ((hours / 24) as u16, (hours % 24) as u8);
+    (days, hours, mins, sec, millis, micros)
+}
+
+/// Format string using `days`, `hours`, `minutes`, `seconds`.
+fn format_dhms(days: u16, hours: u8, minutes: u8, seconds: u8) -> String {
+    match (days, hours, minutes, seconds) {
+        (0, 0, 0, s) => format!("{s:2} s"),
+        (0, 0, m, s) => format!("{m:2} m {s:2} s"),
+        (0, h, m, s) => format!("{h:2} h {m:2} m {s:2} s"),
+        (d, h, m, s) => format!("{d:3} d {h:2} h {m:2} m {s:2} s"),
+    }
+}
+
+/// Format string starting with number of days and going bigger from there.
+fn time_starting_with_seconds(duration: &Duration) -> String {
+    let (days, hours, minutes, seconds, _, _) = get_duration_parts(duration);
+    format_dhms(days, hours, minutes, seconds)
+}
+
+/// Format string starting with number of milliseconds and going bigger from there.
+/// `precision` is how many digits of microseconds from left will be preserved.
+/// Note that this function will always remove all trailing zeros for microseconds.
+fn time_starting_with_ms_with_precision(duration: &Duration, precision: usize) -> String {
+    let digits = precision;
+    let (d, h, m, s, ms, mcs) = get_duration_parts(duration);
+    let mcs_digits = {
+        let mut mcs = mcs;
+        println!("before {mcs}");
+        while mcs % 10 == 0 && mcs != 0 {
+            mcs /= 10
+        }
+        println!("after {mcs}");
+        mcs.to_string().len()
+    };
+    let ms = format!(
+        "{1:.0$} ms",
+        if mcs != 0 { min(digits, mcs_digits) } else { 0 },
+        ms as f32 + mcs as f32 / 1000.0
+    );
+    match (d, h, m, s) {
+        (0, 0, 0, 0) => ms.to_string(),
+        (d, h, m, s) => format_dhms(d, h, m, s),
+    }
+}
+
+fn time_as_ms(duration: &Duration) -> String {
+    time_starting_with_ms_with_precision(duration, 2)
+}
```

Comment:
This can be inlined.

Reply:
Like with `#[inline]` or the wrapper is unnecessary? Note that it's easier when all precisions are set in one compact file and only the wrappers without precision argument are used throughout the codebase. This way, you only have one place to look for if precision should be changed.

Reply:
I meant that the wrapper is unnecessary, yeah.

I'm not sure I agree with yourargument, given the function is only used once. No need for complexity ahead of time without necessity.

If the precision argument shouldn't be exposed, then `time_starting_with_ms_with_precision` shouldn't have the precision argument in the first place, as it is always 2.

Reply:
> If the precision argument shouldn't be exposed, then `time_starting_with_ms_with_precision` shouldn't have the precision argument in the first place, as it is always 2.

But then the name of the function should be changed. Do I just hard-code the precision inside the function?

Reply:
Yeah. Feel free to rename it.